plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/ast/Type.kt

```diff
@@ -6,63 +6,130 @@
+        override fun isSubtype(other: Type): Boolean = other == this
     }
 
     data object Wand : Type {
         override fun toViper(): viper.silver.ast.Type = `Wand$`.`MODULE$`
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Wand = Wand
+        override fun isSubtype(other: Type): Boolean = other == this || other is Bool
     }
 
-    data class Seq(val type: Type) : Type {
-        override fun toViper(): viper.silver.ast.Type = SeqType.apply(type.toViper())
+    data class Seq(val elemType: Type) : Type {
+        override fun toViper(): viper.silver.ast.Type = SeqType.apply(elemType.toViper())
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Seq = Seq(elemType.substitute(typeVarMap))
+        override fun isSubtype(other: Type): Boolean = other is Seq && elemType.isSubtype(other.elemType)
     }
 
-    data class Set(val type: Type) : Type {
-        override fun toViper(): viper.silver.ast.Type = SetType.apply(type.toViper())
+    data class Set(val elemType: Type) : Type {
+        override fun toViper(): viper.silver.ast.Type = SetType.apply(elemType.toViper())
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Set = Set(elemType.substitute(typeVarMap))
+        override fun isSubtype(other: Type): Boolean = other is Set && elemType.isSubtype(other.elemType)
     }
 
-    data class Multiset(val type: Type) : Type {
-        override fun toViper(): viper.silver.ast.Type = MultisetType.apply(type.toViper())
+    data class Multiset(val elemType: Type) : Type {
+        override fun toViper(): viper.silver.ast.Type = MultisetType.apply(elemType.toViper())
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Multiset = Multiset(elemType.substitute(typeVarMap))
+        override fun isSubtype(other: Type): Boolean = other is Multiset && elemType.isSubtype(other.elemType)
     }
 
     data class Map(val keyType: Type, val valueType: Type) : Type {
         override fun toViper(): viper.silver.ast.Type = MapType.apply(keyType.toViper(), valueType.toViper())
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Map =
+            Map(keyType.substitute(typeVarMap), valueType.substitute(typeVarMap))
+
+        override fun isSubtype(other: Type): Boolean = other is Map && keyType.isSubtype(other.keyType) && valueType.isSubtype(other.valueType)
     }
 
     data class TypeVar(val name: String) : Type {
         override fun toViper(): viper.silver.ast.TypeVar =
             viper.silver.ast.TypeVar(name)
+
+        override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>) = typeVarMap.getOrDefault(this, this)
+        override fun isSubtype(other: Type): Boolean = other == this
     }
 
+    // Don't use this type unless defining custom domains.
+    // For expressions see the custom types below which are easier to use.
```

Comment:
I'm a bit confused about the advice here.  When defining a custom domain, when should one define an additional object deriving from `Type` like you do below?